talk/morelanguage/exceptions.tex

```diff
@@ -4,185 +4,172 @@
+      \item throw an exception using {\it throw} when a function cannot proceed (or recover internally)
     \end{itemize}
   \end{block}
   \begin{multicols}{2}
     \begin{cppcode*}{fontsize=\scriptsize,gobble=2}
-      struct ZeroDivide {};
-
-      int divide(int a, int b) {
-        if (0 == b) {
-          throw ZeroDivide();
-        }
-        return a/b;
+      #include <stdexcept>
+      ...
+      try {
+        process_stream_data(s);
+      } catch (const range_error& e) {
+        cerr << e.what() << endl;
       }
     \end{cppcode*}
     \columnbreak
-    \begin{cppcode*}{fontsize=\scriptsize,gobble=2,firstnumber=9}
-      int func(char* value) {
-        try {
-          errno = 0;
-          long l = strtol(value,0,10);
-          if (errno) {
-            throw string("Bad Value");
-          }
-          divide(100, l);
-        } catch (string& e) {
-          printf("%s\n", e.c_str());
-        } catch (ZeroDivide e2) {
-          printf("Division error\n");
+    \begin{cppcode*}{fontsize=\scriptsize,gobble=2}
+      process_stream_data(stream &s) {
+        ...
+        if (data_location >= buffer.length()) {
+          raise range_error("buf overflow");
         }
+        ...
       }
     \end{cppcode*}
   \end{multicols}
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[98]{Controlling exceptions \hfill \deprecated/\removed}
-  \begin{block}{Declaring expected exceptions}
+  \frametitlecpp[98]{Exceptions}
+  \begin{block}{Rules and Advice}
     \begin{itemize}
-    \item each function can declare a set of expected exceptions
-    \item using the {\it throw} statement in its declaration
-    \item other exceptions won't exit the scope of the function
+      \item any object can be thrown; best to use those in \texttt{stdexcept}
+      \begin{itemize}
+        \item define your own subclass of \texttt{std::exception} if needed
+      \end{itemize}
+      \item exceptions will be caught if the object type in the \textit{catch} clause matches
       \begin{itemize}
-      \item instead, the {\it unexpected} handler is called
-      \item by default, it terminates the program
+          \item if nothing catches an exception then \texttt{std::terminate} is called
+          \item throw exceptions by value, catch them by (const) reference
       \end{itemize}
+      \item all objects on the stack between the \textit{throw} and the \textit{catch} are destructed automatically
+      \begin{itemize}
+        \item this should give you a clean release of intermediate resources
+        \item make sure you are using the RAII idiom in your own classes
+      \end{itemize}
+      \item use exceptions for \textit{unlikely} runtime errors outside the program's control (bad inputs, files unexpectedly not found, DB connection, \ldots)
+      \item \textit{don't} use exceptions for logic errors in your code (consider \texttt{assert} and tests)
```

Comment:
Maybe add:
don't use exceptions to provide no or an alternative return value. Use `std::optional`, `std::variant` or `boost::outcome` (or `std::expected` in C++23)`

Reply:
Is it actually ok to suggest returning a custom struct/pair of outval,error_code as an alternative as well (and also mention it is not preferred to manipulate global errno like C)

Reply:
I split this slide in two now to make it more complete on these points (it's a bit wordy, but a lot more informative)

Reply:
It would seem that `std::optional` and `std::variant` are better than the older `std::pair` formalism, right?